### PR TITLE
Visualization-239-john-s-f PR

### PR DESCRIPTION
**Pull request by FlicAnderson**
This pull request should solve #239, adding interactive visualization code to riboviz from branch `visualization-239-john-s-f`. 

This PR code: 
* generates 1 new output file to `dir_in`: `interactive_viz_config.yaml`. 
* adds 1 new script to `rscripts`: `run_shiny_server.R`. 
* adds 1 new documentation file to `docs/user/`: `run-shiny-server-operation.md` 
* updates package install list in `docs.user/install.md`.

**Requires Action**: 

Needs installation of new R packages: 
```
install.packages(scales)  
install.packages(shiny)
install.packages(yaml)
```  

Requesting review by @mikej888, @swinterbourne and @3mma-mack, covering roughly the areas [mentioned here](https://github.com/riboviz/riboviz/issues/239#issuecomment-901263692)

### CODE REVIEW

**Comment by ewallace** [comment]
I don't remember how Eddie deals with R packages, but do we need to update `set-riboviz-env.sh` or any of the R versions / packages there????

In other words, @3mma-mack and @swinterbourne, if you ran this on Eddie, what did you have to change to do so?

**Comment by mikej888** [comment]
`set-riboviz-env.sh` doesn't do anything R-specific - it only sets the PATH for hisat2 and bowtie and activates the `riboviz` Miniconda environment.

**Review by 3mma-mack — COMMENTED** [review]
I have tried running the vignette, producing the interactive config file on Eddie and then downloading the output files to my local computer and running with the command ``

This worked successfully, producing the expected vizualisation of plots. I also no longer ran into the error:` Error in h5checktypeOrOpenLoc(). Cannot open file. File 'C:\Users\emma\Git\riboviz\B-Sc_2012\output\C:\Users\emma\Git\riboviz\B-Sc_2012\output\VEG_1\C:\Users\emma\Git\riboviz\B-Sc_2012\output\VEG_1.h5' does not exist
`, but I will double check this by testing the Brar dataset this afternoon or tomorrow.

I did not have an issue with any R packages, as the new packages are called in run_shiny_server.R, which I run on my computer and I have previously installed the packages. 
There also did not seem to be a problem when run_shiny_server.R was run on eddie, as it did produce `Listening on http://127.0.0.1:4254`, however I would need to learn how to forward the port to know for sure.

With that in mind I definitely agree with the line from docs/user/run-shiny-server-operation.md `A description of how to forward the port from the command line as opposed to doing it in VSC would be good here...`. 

But overall it is really good, and works to produce the expected outputs.

File: docs/user/prep-riboviz-operation.md [hunk 1]
```diff
@@ -21,7 +21,15 @@ The workflow prepares ribosome profiling data for by implementing a workflow tha
 * `bam_to_h5.R`: convert BAM to compressed H5 format (local script, in `rscripts/`)
 * `generate_stats_figs.R`: generate summary statistics, analyses plots and QC plots (local script, in `rscripts/`)
 * `collate_tpms.R`: collate TPMs across samples (local script, in `rscripts/`)
-* `riboviz.tools.count_reads`: count the number of reads (sequences) processed by specific stages of the workflow (local script, in `riboviz/tools/`).
+* `riboviz.tools.count_reads`: count the number of reads (sequences) processed by specific stages of the workflow (local script, in `riboviz/tools/`).  
+* `AnalysisOutputs.Rmd`: the `staticHTML` Nextflow process runs an R markdown document which generates an HTML output report for each sample processed, using the analysed data (local script, in `rmarkdown/`)
+
+#### Interactive Visualization
```
Comment (mikej888) [inline docs/user/prep-riboviz-operation.md:27]: Change to level 3 heading, `###`, as surrounding section is level 2, `##`.

File: docs/user/prep-riboviz-operation.md [hunk 2]
```diff
@@ -21,7 +21,15 @@ The workflow prepares ribosome profiling data for by implementing a workflow tha
...
+
+An additional script within the riboviz codebase allows the generation of interactive plots from the dataset after the riboviz workflow has been completed.
```
Comment (mikej888) [inline docs/user/prep-riboviz-operation.md:29]: riboviz => RiboViz

File: docs/user/prep-riboviz-operation.md [hunk 3]
```diff
@@ -21,7 +21,15 @@ The workflow prepares ribosome profiling data for by implementing a workflow tha
...
+* `riboviz.tools.count_reads`: count the number of reads (sequences) processed by specific stages of the workflow (local script, in `riboviz/tools/`).  
+* `AnalysisOutputs.Rmd`: the `staticHTML` Nextflow process runs an R markdown document which generates an HTML output report for each sample processed, using the analysed data (local script, in `rmarkdown/`)
+
+#### Interactive Visualization
+
+An additional script within the riboviz codebase allows the generation of interactive plots from the dataset after the riboviz workflow has been completed.  
+
+A process within the main riboviz workflow (`createInteractiveVizParamsConfigFile`, within `prep-riboviz.nf`) generates a configuration yaml file specifically for this interactive visualization (`interactive_viz_config.yaml`, found in the `dir_out` directory), and users can then run `rscripts/run_shiny_server.R` using this additional yaml (as detailed within the specific [run-shiny-server-operation.md documentation](./run-run_shiny_server-operation.md)).  This generates a shiny app instance which users can view to explore their data.
```
Comment (mikej888) [inline docs/user/prep-riboviz-operation.md:31]: riboviz => RiboViz 

```
as detailed within the specific [run-shiny-server-operation.md documentation](./run-run_shiny_server-operation.md)
```
=>
```
as detailed within [How To Run the Riboviz Interactive Data Visualization On Your Data](./run-run_shiny_server-operation.md)
```
 i.e. page title not page file name.

yaml => YAML

File: docs/user/prep-riboviz-operation.md [hunk 4]
```diff
@@ -159,7 +167,8 @@ For each sample (`<SAMPLE_ID>`), intermediate files are produced in a sample-spe
 In addition, the following files are also put into the output directory:
 
 * `TPMs_collated.tsv`: file with the transcripts per million (tpm) for all successfully processed samples.
-* `read_counts.tsv`: a [read counts file](#read-counts-file) (only if `count_reads: TRUE`).
+* `read_counts.tsv`: a [read counts file](#read-counts-file) (only if `count_reads: TRUE`).  
+* `interactive_viz_config.yaml`: this is a yaml file created by the workflow, for use with `rscripts/run_shiny_server.R` - an optional step which does not automatically run within the riboviz workflow and which allows users to generate interactive data visualization on the dataset.
```
Comment (mikej888) [inline docs/user/prep-riboviz-operation.md:171]: riboviz => RiboViz

File: docs/user/run-shiny-server-operation.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
+# How To Run the Riboviz Interactive Data Visualization On Your Data
```
Comment (mikej888) [inline docs/user/run-shiny-server-operation.md:1]: riboviz => RiboViz throughout the page

File: docs/user/run-shiny-server-operation.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+This page describes how to generate and run an interactive shiny app to visualize and explore your data after running the riboviz workflow. Riboviz 2.0 has an integrated, but optional data visualization tool written in `R` and using `shiny`. The script is located at `riboviz/rscripts/run_shiny_server.R` and takes a single mandatory argument - a YAML file, and one optional argument - a port number for port forwarding. The YAML file is not the one used for a riboviz run, but instead is one called `interactive_viz_config.yaml` which will be located in your output folder.
```
Comment (mikej888) [inline docs/user/run-shiny-server-operation.md:3]: Remove mention of "2.0". This feature now applies to all releases from the current one.

Comment (mikej888) [reply docs/user/run-shiny-server-operation.md:3]: Rename `riboviz/rscripts/run_shiny_server.R` to `rscripts/run_shiny_server.R`. Keep references to file paths within the scope of the RiboViz repository except where necessary (e.g. if referring to absolute paths).

File: docs/user/run-shiny-server-operation.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+# Activating the shiny server on a local machine
```
Comment (mikej888) [inline docs/user/run-shiny-server-operation.md:5]: Update level 1, `##`, subheadings to level 2, `##`.

File: docs/user/run-shiny-server-operation.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+If you're running on a headless machine like a server, the shiny server will be hosted on said server. By default, the shiny server is sent to port `4254`. If this port is open, then the same command as above will allow you to access the shiny server. Alternatively, you can specify a port at the command line as in the following example which uses port `1234`.
+```
+Rscript --vanilla run_shiny_server.R path/to/yaml/configuration/file.yaml 1234
+```
+
+You may need to forward that port. *A description of how to forward the port from the command line as opposed to doing it in VSC would be good here...*
+
+# Description of tabs
+
+Globally, the user can choose which samples they would like to view by checking the appropriate boxes at the top of the screen. Changes are not submitted until the user presses the `apply` button.
```
Comment (mikej888) [inline docs/user/run-shiny-server-operation.md:30]: 'apply' button => 'Apply Changes' button

File: docs/user/run-shiny-server-operation.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+```
+Rscript --vanilla run_shiny_server.R path/to/interactive_viz_config.yaml
+```
+
+Once entered, the command will print out the something like the following
+```
+"Listening on http://127.0.0.1:4254"
+```
+
+The IP address provided can simply be pasted into a web browsers URL bar and will direct you to the shiny server which is now hosted from the local machine.
```
Comment (mikej888) [inline docs/user/run-shiny-server-operation.md:17]: browsers => browser's

File: docs/user/run-shiny-server-operation.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+### Read length distributions
+
+This tab displays read length distributions.
+
+### 3nt periodicity
+
+This tab displays graphs related to the positioning and the frame of the reads.
+
+### Normalized reads per position
```
Comment (mikej888) [inline docs/user/run-shiny-server-operation.md:48]: Normalized reads per position => Normalized reads by position

File: docs/user/run-shiny-server-operation.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+The IP address provided can simply be pasted into a web browsers URL bar and will direct you to the shiny server which is now hosted from the local machine. 
+
+# Activating the shiny server on a headless machine
+
+If you're running on a headless machine like a server, the shiny server will be hosted on said server. By default, the shiny server is sent to port `4254`. If this port is open, then the same command as above will allow you to access the shiny server. Alternatively, you can specify a port at the command line as in the following example which uses port `1234`.
+```
+Rscript --vanilla run_shiny_server.R path/to/yaml/configuration/file.yaml 1234
+```
+
+You may need to forward that port. *A description of how to forward the port from the command line as opposed to doing it in VSC would be good here...*
```
Comment (mikej888) [inline docs/user/run-shiny-server-operation.md:26]: * Incomplete text. Either delete this comment or provide a how-to or a web link.

**Comment by mikej888** [comment]
`shiny` is already a documented dependency. `scales` and `yaml` were also already installed as side-effects of other R dependencies but it is good to record these explicitly.

Edit `README.md` and add bullet with link to "Usage:", under "Running the RiboViz Nextflow workflow": `[How To Run the Riboviz Interactive Data Visualization On Your Data](docs/user/run-run_shiny_server-operation.md)`

Edit `rscripts/install-r.R` and add `scales` and `yaml`.

**Comment by mikej888** [comment]
There is no integration test to check for the existence of `interactive_viz_config.yaml`. Is this intentional?

Usability issues:

* If the user changes the value of "Gene" in the "TPM summary" tab to "xxx" and clicks "Apply Changes" the graphs do not change. Is something like a "gene not found" error being silently caught and ignored?
* If the user changes the value of "Gene" in the "Features" tab to "xxx" and clicks "Apply Changes" the graphs do not change. As for the above.
* If the user then goes to the "Gene specific coverage" tab, "Gene" shows "YAL001C" but there is an error message "Object 'xxx/vignette/reads/data' does not exist in this HDF5 file.". Is the "Gene" field displaying "Gene" for the wrong tab?

**Comment by mikej888** [comment]
Shiny server ran OK under my Ubuntu 18.04, Ubuntu 20.04.2, CentOS 7 2009 environments:
```console
$ nextflow run prep_riboviz.nf -ansi-log false -params-file vignette/vignette_config.yaml
$ Rscript --vanilla rscripts/run_shiny_server.R vignette/output/interactive_viz_config.yaml
```
The multiplexed example also worked OK under Ubuntu 18.04:
```console
$ nextflow run prep_riboviz.nf -ansi-log false -params-file vignette/simdata_multiplex_config.yaml 
$ Rscript --vanilla rscripts/run_shiny_server.R vignette/simdata_multiplex_output/interactive_viz_config.yaml 
```

**Comment by mikej888** [comment]
For `rscripts/run_shiny_server.R` apply R coding style as in `docs/developer/coding-style.md` including addressing `lintr` recommendations.

File: docs/user/prep-riboviz-operation.md [hunk 12]
```diff
@@ -44,7 +52,7 @@ If sample files (`fq_files`) are specified, then the workflow processes the samp
    10. Write intermediate files produced above into a sample-specific directory, named using the sample ID, within the temporary directory (`dir_tmp`).
    11. Make length-sensitive alignments in compressed h5 format using `bam_to_h5.R`.
    12. Generate summary statistics, and analyses and QC plots for both RPF and mRNA datasets using `generate_stats_figs.R`. This includes estimated read counts, reads per base, and transcripts per million for each ORF in each sample.
-   13. Write output files produced above into an sample-specific directory, named using the sample ID, within the output directory (`dir_out`). 
+   13. Write output files produced above into an sample-specific directory, named using the sample ID, within the output directory (`dir_out`).
```
Comment (swinterbourne) [inline docs/user/prep-riboviz-operation.md:55]: Replace "an" with "a" in `Write output files produced above into an sample-specific directory`?

File: docs/user/run-shiny-server-operation.md [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+Globally, the user can choose which samples they would like to view by checking the appropriate boxes at the top of the screen. Changes are not submitted until the user presses the `apply` button.
+
+### Read count summary
+
+This tab displays a bar graph showing the number reads over the course of the processing steps.
+
+### TPM summary
+
+This tab display a heatmap with sample to sample pairwise correlations based on $log_{10}(TPM)$ as well as TPM density plots showing the distribution of TPMs for each gene. This tab allows you to enter the name of gene from your data, doing so will draw a vertical black line on the density plots to show you where that gene lies among the distribution of TPMs.
```
Comment (swinterbourne) [inline docs/user/run-shiny-server-operation.md:38]: display => displays

For the second sentence - maybe insert "the" here: `This tab allows you to enter the name of <the> gene from your data` ?

**Comment by mikej888** [comment]
There is a bug if the user provides their own port, even the default:
```console
$ Rscript rscripts/run_shiny_server.R vignette/output/interactive_viz_config.yaml  4254
Listening on domain socket 4254
Error in startApp(appParts, port, host, quiet) : 
  `port` is not a valid domain socket (missing `mask` attribute). Note that if you're using the default `host` + `port` configuration (and not domain sockets), then `port` must be numeric, not a string.
Calls: <Anonymous> ... print.shiny.appobj -> do.call -> runApp -> startApp
Execution halted
```
This is likely due to the `args` being `character` values and the line
```
  port_num <- args[2]
```
which would result in `port_num` being a `character`. It needs parsed to an integer using `strtoi`.

**Comment by mikej888** [comment]
I had a go at port forwarding and got a couple of examples working using a "hello world" example from [Your first Shiny app](https://mastering-shiny.org/basic-app.html):
```R
library(shiny)
port_num <- 4254
ui <- fluidPage(
  "Hello, world!"
)
server <- function(input, output, session) {
}
shinyApp(ui, server, options = list(port = port_num))
```
I got a simple example working, where a remote host is running the Shiny server, and an Eddie example, where one has to tunnel through a login node to an interactive node.

Suggested content for the "Activating the shiny server on a headless machine" is below. It would be useful if someone could check each of these using the actual `Rscript --vanilla rscripts/run_shiny_server.R vignette/output/interactive_viz_config.yaml` command.

---

Here are a couple of examples of using port forwarding to access the Shiny server when it is run on a machine that does not allow the use of browsers. For more information see, for example, [SSH port forwarding - Example, command, server config](https://www.ssh.com/academy/ssh/tunneling/example#remote-forwarding).

### Port forwarding to a remote host

Within the host in which you ran RiboViz, start the Shiny server. For example:

```console
$ Rscript --vanilla rscripts/run_shiny_server.R vignette/output/interactive_viz_config.yaml
```

On your local machine, create an SSH tunnel to the host where you are running the Shiny server. For example:

```console
$ ssh -NL localhost:4254:localhost:4254 user@somehost.ac.uk
```

* `-N` instructs `ssh` to not execute a remote command i.e. forward connections only.
* `-L localhost:4254:localhost:4254` instructs `ssh` to forward connections to `localhost:4254` on your local machine to `localhost:4254` on the remote host i.e. to the Shiny server you started on the remote host.

On your local machine, in a browser, visit http://localhost:4254. You should see the Shiny visualisations.

### Port forwarding and Eddie

Log into Eddie and get name of login node you are connected to. For example:

```console
$ ssh <USER>@eddie.ecdf.ed.ac.uk
$ hostname
login01.ecdf.ed.ac.uk
```

Create an interactive session. For example:

```console
$ qlogin
```

From the interactive node, create an SSH tunnel back to the specific login node you were connected to. For example:

```console
$ ssh -NR :4254:localhost:4254 login01 &
```

* `-N` instructs `ssh` to not execute a remote command i.e. forward connections only.
* `-R :4254:localhost:4254` instructs `ssh` forward connections to 4254 on the login node to `localhost:4254` on the interactive node.

Within the interactive node, start the Shiny server. For example:

```console
$ source set_riboviz_env.sh
$ Rscript --vanilla rscripts/run_shiny_server.R vignette/output/interactive_viz_config.yaml
```

In a new terminal window on your local machine, create an SSH tunnel to the specific Eddie login node you are connected to, adding an `-ext` suffix to the login node name. For example:

```console
$ ssh -NL localhost:4254:localhost:4254 <USER>@login01-ext.ecdf.ed.ac.uk
```

* `-N` instructs `ssh` to not execute a remote command i.e. forward connections only.
* `-L localhost:4254:localhost:4254` instructs `ssh` to forward connections to `localhost:4254` on your local machine to `localhost:4254` on the remote host the login node. There they will be forwarded, using the `ssh` session you created on the interactive node, on to `localhost:4254` on the interactive node i.e. to the Shiny server.

On your local machine, in a browser, visit http://localhost:4254. You should see the Shiny visualisations.

(these instructions were adapted from [Jupyter notebooks on Eddie](https://www.wiki.ed.ac.uk/display/geoinf/Jupyter+notebooks+on+Eddie))

File: rscripts/run_shiny_server.R [hunk 14]
```diff
@@ -0,0 +1,755 @@
...
+                    )
+                  }),
+                  headerPanel(""),
+                  plotOutput("features_plot", height = "1000px")
+                )
+              )
+            )))
+
+# I cast spell, ~run shiny app~
+shinyApp(ui, server, options = list(port = port_num))
```
Comment (3mma-mack) [inline rscripts/run_shiny_server.R:755]: Is it a problem if there is no new line at the end of a file? It seems to be flagged with a little stop sign so I thought I'd call attention to it.

File: rscripts/run_shiny_server.R [hunk 15]
```diff
@@ -0,0 +1,755 @@
...
+### define the universal plot theme
+plot_theme <- theme_bw() +
+  theme(
+    panel.background = element_blank(),
+    panel.grid = element_blank(),
+    text = element_text(size = 15)
+  )
+
+### Start the server
+server <- function(input, output, session) {
```
Comment (3mma-mack) [inline rscripts/run_shiny_server.R:204]: Potentially change the name of this function to be more descriptive? I think this function is called on line 755 with line 
`shinyApp(ui, server, options = list(port = port_num))` but I am not entirely sure, so a clearer name would clarify this.

server => create_server

Edit: **I have looked at R documentation for shinyApp and it requires a function called server, so what you have here is fine, I am just unfamiliar with `shiny`

**Review by 3mma-mack — COMMENTED** [review]
Having looked further into rscripts/run_shiny_server.R I really liked the style of the code, it was very clear what was being done in each section, and the comments were very helpful. I couldn't see any obvious mistakes or places that could lead to bugs.

File: docs/user/prep-riboviz-operation.md (hunk 2 above)
Comment (FlicAnderson) [reply docs/user/prep-riboviz-operation.md:29]: Sorry, I think I thought we were using lower case riboviz capitalization throughout?

File: docs/user/prep-riboviz-operation.md (hunk 3 above)
Comment (FlicAnderson) [reply docs/user/prep-riboviz-operation.md:31]: Good catches, thanks

File: docs/user/run-shiny-server-operation.md (hunk 8 above)
Comment (FlicAnderson) [reply docs/user/run-shiny-server-operation.md:30]: Agreed, will change

File: docs/user/run-shiny-server-operation.md (hunk 11 above)
Comment (FlicAnderson) [reply docs/user/run-shiny-server-operation.md:26]: I'll have a short google for this and if I can't find any methods that seem to be quick and easy, I'll remove the comment and create a new ticket for this

File: docs/user/prep-riboviz-operation.md (hunk 12 above)
Comment (FlicAnderson) [reply docs/user/prep-riboviz-operation.md:55]: Good catch :+1:

**Comment by FlicAnderson** [comment]
Thanks for all these comments/feedback so far folks @mikej888 @swinterbourne and @3mma-mack - I'll get these integrated asap!

**Comment by FlicAnderson** [comment]
> Usability issues:
> 
> * If the user changes the value of "Gene" in the "TPM summary" tab to "xxx" and clicks "Apply Changes" the graphs do not change. Is something like a "gene not found" error being silently caught and ignored?
> * If the user changes the value of "Gene" in the "Features" tab to "xxx" and clicks "Apply Changes" the graphs do not change. As for the above.
> * If the user then goes to the "Gene specific coverage" tab, "Gene" shows "YAL001C" but there is an error message "Object 'xxx/vignette/reads/data' does not exist in this HDF5 file.". Is the "Gene" field displaying "Gene" for the wrong tab?

@john-s-f are these issues something you and/or @p6t can comment on at all?  Any ideas how we could resolve these?

**Comment by kavousan** [comment]
Thanks for the findings @FlicAnderson. 

Can folk also opine if any of these hold us back from releasing (and submitting the manuscript)?

**Comment by FlicAnderson** [comment]
> There is a bug if the user provides their own port, even the default:
> 
> ```
> $ Rscript rscripts/run_shiny_server.R vignette/output/interactive_viz_config.yaml  4254
> Listening on domain socket 4254
> Error in startApp(appParts, port, host, quiet) : 
>   `port` is not a valid domain socket (missing `mask` attribute). Note that if you're using the default `host` + `port` configuration (and not domain sockets), then `port` must be numeric, not a string.
> Calls: <Anonymous> ... print.shiny.appobj -> do.call -> runApp -> startApp
> Execution halted
> ```
> 
> This is likely due to the `args` being `character` values and the line
> 
> ```
>   port_num <- args[2]
> ```
> 
> which would result in `port_num` being a `character`. It needs parsed to an integer using `strtoi`.

Hi @mikej888, is there a technical reason why the more 'typical' base R function `as.integer()` wouldn't work here instead?  `strtoi()` seems to be based on a C function equivalent, but isn't very commonly used in R, although it's part of base R.  

From R help for `as.integer()`:
> `as.integer` attempts to coerce its argument to be of integer type. The answer will be NA unless the coercion succeeds. 

Are the particular conversion details using C rules particularly important, or is this just a case that either of these functions would do the job and avoid the bug?

**Comment by ewallace** [comment]
Is port forwarding a separate issue ticket?

**Comment by FlicAnderson** [comment]
> There is no integration test to check for the existence of `interactive_viz_config.yaml`. Is this intentional?

@mikej888  This was not intentional, I think we definitely should have an integration to test for the existence of this yaml file, particularly since it's generated in each run of riboviz!  

I wonder whether in a subsequent issue we should beef up the test to also check the values within the yaml are non-empty, if this is something that's not too complex to do?

**Comment by FlicAnderson** [comment]
Commit 8f5e346c78 updates `rscripts/run_shiny_server.R` to match file names following merged work on #411. 

One file name will need to be updated when PR #428 is merged: 
Currently `gene_position_length_counts_5start.tsv`, will become `metagene_position_length_counts_5start.tsv` (details [here](https://github.com/riboviz/riboviz/issues/426#issuecomment-916068882).  As this PR is not yet currently merged, I haven't updated this in `rscripts/run_shiny_server.R`

**Comment by FlicAnderson** [comment]
The small tweaks, typos, rewordings & such have now been done. 

The shiny code is now up to date with the renamed files following the work on #411.  

Still to do: 

1. add integration test for `interactive_viz_config.yaml`  
2. apply lintr style recommendations to `rscripts/run_shiny_server.R`  
3. either add [Mike's port forwarding information](https://github.com/riboviz/riboviz/pull/417#issuecomment-905429507) to `/docs/user/run-shiny-server-operation.md` documentation, or create new issue to add this subsequently  
4. decide whether to fix shiny app usability issues brought up by Mike [here](https://github.com/riboviz/riboviz/pull/417#issuecomment-903674706) within this PR, or address as separate issues subsequently. 

Personally, I think item 4 could probably be handled separately by creating new issues and following those up there, and shouldn't hold back merging this PR and resubmitting the paper.  Opinions from @john-s-f would be great on this.

@ewallace suggests that perhaps port forwarding (item 3) could be a separate issue, and perhaps it might be quicker to merge this PR to develop and then test and add that documentation as a new issue which would just be a small tweak, and wouldn't hold anything up?  I'd be tempted to add Mike's in-depth port forwarding information (item 3) within this PR, but I would need to try and test it out on Eddie first, which again might slow things down a little for the paper. 

Item 2 should be very quick to implement (being mostly 'remove superfluous whitespace' and 'lines should not be more than 80 characters'). 

Item 1 - I think this shouldn't take too long to do since we just want to check a file exists, but it will take me a while to figure out how to implement this based on the format of existing integration tests, and getting my head around what the process was for previous PRs I've done, which files need changes, etc. @mikej888 if you know off the top of your head of any examples in our existing codebase which should be fairly similar, it'd be great to know, as that would speed up my search time.

**Comment by FlicAnderson** [comment]
As an approximate timeline, I think I can get 1 & 2 sorted and ready to merge by tomorrow's dev meeting (15th Sept 2021). Possibly 3 as well, though I think perhaps handling this as a separate issue might add more certainty to the timings for this PR.

**Comment by FlicAnderson** [comment]
Item 2 (lintr checks & fixes) now done as of commit cb13dde48b. 

Item 1 (integration test check for `interactive_viz_config.yaml`) still to do.

**Comment by FlicAnderson** [comment]
I've created a new issue for item 3 (port forwarding documentation): #431 and therefore won't include this within this PR.

**Comment by mikej888** [comment]
> Hi @mikej888, is there a technical reason why the more 'typical' base R function `as.integer()` wouldn't work here instead? `strtoi()` seems to be based on a C function equivalent, but isn't very commonly used in R, although it's part of base R.
> 
> From R help for `as.integer()`:
> 
> > `as.integer` attempts to coerce its argument to be of integer type. The answer will be NA unless the coercion succeeds.
> 
> Are the particular conversion details using C rules particularly important, or is this just a case that either of these functions would do the job and avoid the bug?

@FlicAnderson I chose `strtooi()` as it was what Google revealed! The [strtoi](https://stat.ethz.ch/R-manual/R-devel/library/base/html/strtoi.html) docs comment that

> For decimal strings as.integer is equally useful.

so either should be fine.

**Comment by mikej888** [comment]
@FlicAnderson I've created visualization-239-john-s-f-mikej888. This has riboviz.test.integration.test_integration with a new `test_interactive_viz_config_yaml` test. It validates the interactive YAML file content against the known configuration [I extended riboviz.test.integration.conftest` to create the additional test fixtures required with `fq_files`, `sample_sheet` etc]. Please have a look and give it a run.

**Comment by mikej888** [comment]
I'd also added `INTERACTIVE_VIZ_CONFIG_FILE` to `riboviz.workflow_files` to my branch which duplicates your constant.

**Comment by FlicAnderson** [comment]
> @FlicAnderson I've created visualization-239-john-s-f-mikej888. This has riboviz.test.integration.test_integration with a new `test_interactive_viz_config_yaml` test. It validates the interactive YAML file content against the known configuration [I extended `riboviz.test.integration.conftest` to create the additional test fixtures required with `fq_files`, `sample_sheet` etc]. Please have a look and give it a run.

@mikej888 legend! I was working very slowly and tentatively on doing this, but hadn't got very far (had created a rough structure for a new test in test_integration.py only).  

I'm currently checking & running your tests. Will report back.

**Comment by FlicAnderson** [comment]
@mikej888 this works a treat - pinged up a fail when I deleted the `interactive_viz_config.yaml` file from `dir_out` and ran the tests using `--skip-workflow`, and when I put it back and deleted a parameter from the file, I got a fail letting me know which parameter was missing.  This looks great. 

I've merged in `visualization-239-john-s-f-mikej888` to branch `visualization-239-john-s-f`, fixed the merge conflict by binning my constant `INTERACTIVE_VIZ_YAML` and choosing Mike's one. 

The latest `develop` is merged in, and I think it's pretty much good to go now?

File: docs/user/run-shiny-server-operation.md (hunk 11 above)
Comment (FlicAnderson) [reply docs/user/run-shiny-server-operation.md]: This issue has been split out as a new ticket: [riboviz/#431](https://github.com/riboviz/riboviz/issues/431), so I'm marking it as resolved here.

**Comment by FlicAnderson** [comment]
I've pulled out @mikej888's usability issues relating to the shiny app UI and created a new issue for these (#433).

**Comment by mikej888** [comment]
For #431, see [comment](https://github.com/riboviz/riboviz/issues/431#issuecomment-920077251) and decide whether you're happy to merge in the doc changes I've made.

**Comment by mikej888** [comment]
@FlicAnderson, in `rscripts/run_shiny_server.R`:
```
  port_num <- args[as.integer(2)]  # use strtoi() to convert to int, avoids bug
```
needs to be
```
  port_num <- as.integer(args[2])
```
The former converts `2` to an integer then uses that as an index into `args`.

**Comment by FlicAnderson** [comment]
> @FlicAnderson, for #431, see [comment](https://github.com/riboviz/riboviz/issues/431#issuecomment-920077251) and decide whether you're happy to merge in the doc changes I've made.

I haven't been able to check this quickly before the dev meeting due to a git tangle on Eddie, but I think it's probably worth merging it in & we can adapt the wording in future if there's any usability feedback from users.  Can discuss in meeting?

**Comment by mikej888** [comment]
@FlicAnderson I think you should also have the pleasure of merging `develop` into `main`!

**Comment by FlicAnderson** [comment]
> @FlicAnderson I think you should also have the pleasure of merging `develop` into `main`!

@mikej888  Is this done via a new PR?  What's the procedure?  I've never done that before... :astonished:

**Comment by mikej888** [comment]
@FlicAnderson, yes, raise a new PR of `develop` against `main` then merge it.

**Pull request merged**